servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancer.java

```diff
@@ -331,64 +337,77 @@ public Completable closeAsyncGracefully() {
-            for (C conn : toRemove) {
-                conn.closeAsyncGracefully().subscribe();
+            final Object[] toRemove = connectionsUpdater.getAndSet(this, CLOSED_ARRAY);
+            LOGGER.debug("Closing {} connection(s) gracefully to inactive address: {}", toRemove.length, address);
+            for (Object conn : toRemove) {
+                @SuppressWarnings("unchecked")
+                final C cConn = (C) conn;
+                cConn.closeAsyncGracefully().subscribe();
             }
         }
 
         boolean isInactive() {
-            return connections == CLOSED_LIST;
+            return connections == CLOSED_ARRAY;
         }
 
         boolean addConnection(C connection) {
             for (;;) {
-                List<C> existing = this.connections;
-                if (existing == CLOSED_LIST) {
+                final Object[] existing = this.connections;
+                if (existing == CLOSED_ARRAY) {
                     return false;
                 }
-                ArrayList<C> connectionAdded = new ArrayList<>(existing);
-                connectionAdded.add(connection);
-                if (connectionsUpdater.compareAndSet(this, existing, connectionAdded)) {
+                Object[] newList = Arrays.copyOf(existing, existing.length + 1);
+                newList[existing.length] = connection;
+                if (connectionsUpdater.compareAndSet(this, existing, newList)) {
                     break;
                 }
             }
 
             // Instrument the new connection so we prune it on close
             connection.onClose().beforeFinally(() -> {
                 for (;;) {
-                    final List<C> existing = connections;
-                    if (existing == CLOSED_LIST) {
+                    final Object[] existing = this.connections;
+                    if (existing == CLOSED_ARRAY) {
                         break;
                     }
-                    ArrayList<C> connectionRemoved = new ArrayList<>(existing);
-                    if (!connectionRemoved.remove(connection) ||
-                            connectionsUpdater.compareAndSet(this, existing, connectionRemoved)) {
+                    int i = 0;
+                    for (; i < existing.length; ++i) {
+                        if (existing[i].equals(connection)) {
+                            break;
+                        }
+                    }
+                    if (i == existing.length) {
                         break;
+                    } else {
+                        Object[] newList = new Object[existing.length - 1];
+                        System.arraycopy(existing, 0, newList, 0, i);
+                        System.arraycopy(existing, i + 1, newList, i, newList.length - i);
+                        if (connectionsUpdater.compareAndSet(this, existing, newList)) {
+                            break;
+                        }
                     }
                 }
             }).subscribe();
```

Comment:
Perhaps make this a standalone removeConnectionMethod() and call that in beforeFinally?